cryptoki/src/functions/session_management.rs

```diff
@@ -28,8 +27,93 @@ impl Pkcs11 {
             ))
             .into_result()?;
         }
+        Ok(session_handle)
+    }
+
+    /// Open a new session with no callback set
+    pub fn open_session_no_callback(&self, slot_id: Slot, flags: SessionFlags) -> Result<Session> {
+        Ok(Session::new(self.open_session(slot_id, flags)?, self))
+    }
+
+    /// Opens a new management session on a token
+    ///
+    /// The purpose of a management session is to hold the state of all sessions opened on a token
+    /// by the running application.  This allows threads to open their own session, already at the
+    /// desired state set by the management session.
+    ///
+    /// # Arguments
+    ///
+    /// * `slot_id` - The slot to open the session on
+    /// * `flags` - The flags for the session
+    /// * `user_type` - The type of user to login
+    /// * `pin` - The PIN to use, or `None` if a separate authentication path exists
+    ///
+    /// # Errors
+    ///
+    /// If the session fails to open, the typical [`Error::Pkcs11`] will return.  If a management
+    /// session already exists for the token, then [`Error::ManagementSessionExists`] will be
+    /// returned.
+    pub fn open_management_session(
+        &self,
+        slot_id: Slot,
+        flags: SessionFlags,
+        user_type: UserType,
+        pin: Option<&str>,
+    ) -> Result<()> {
+        let mut map = self
+            .management_sessions
+            .lock()
+            .map_err(|_e| crate::Error::MutexPoisonError)?;
+        if let std::collections::hash_map::Entry::Vacant(_) = map.entry(slot_id) {
+            let session_handle = self.open_session(slot_id, flags)?;
+            let (pin, pin_len) = match pin {
+                Some(pin) => (pin.as_ptr() as *mut u8, pin.len()),
+                None => (std::ptr::null_mut(), 0),
+            };
+            unsafe {
+                Rv::from(get_pkcs11!(self, C_Login)(
+                    session_handle,
+                    user_type.into(),
+                    pin,
+                    pin_len.try_into()?,
+                ))
+                .into_result()
+                .map_err(|e| {
+                    // Close the session if we can't login and return the resulting error
+                    // Not using `get_pkcs11` here because the macro uses a '?' which is not
+                    // allowed in a closure that does not return a Result.
+                    // Justification for unwrap -> If C_CloseSession doesn't exist, the
+                    // library is frankly unusable.
+                    let _ = self.function_list.C_CloseSession.unwrap()(session_handle);
```

Comment:
I think even if the library is unusable we shouldn't bomb out. Maybe something like:
```rust
self.function_list.C_CloseSession.ok_or(Error::NullFunctionPointer)?(session_handle);
```
?

Reply:
The problem is that you can't use '?' in this closure because the closure returns an `Error` not a `Result`.

Maybe a better solution would be to log the error (which I personally don't like because I don't like libraries logging anything unless explicitly told to do so) or maybe just ignore it?

Reply:
ah, yes, that's true, and it might not be desirable to return that error instead of the "original" one anyway. Just ignore it if `C_CloseSession` doesn't exist